Add method for cancelling peripheral connections

jonathanhogg opened the pull request:
Add a new `gap_connect_cancel()` method to allow an in-progress `gap_connect()` to be cancelled. This allows for timing-out a stuck connection attempt – typical of unreliable links or devices that suddenly disappear.

Tested on the ESP32 nimble port, but untested on btstack.

jimmo commented on extmod/nimble/modbluetooth_nimble.c:
```diff
@@ -949,6 +949,16 @@ int mp_bluetooth_gap_peripheral_connect(uint8_t addr_type, const uint8_t *addr,
+    }
+    int err = ble_gap_conn_cancel();
+    return ble_hs_err_to_errno(err);
+
```
Extra line

jimmo commented on extmod/nimble/modbluetooth_nimble.c:
```diff
@@ -949,6 +949,16 @@ int mp_bluetooth_gap_peripheral_connect(uint8_t addr_type, const uint8_t *addr,
     return ble_hs_err_to_errno(err);
 }
 
+int mp_bluetooth_gap_peripheral_connect_cancel() {
```
Should be `(void)`  (also in modbluetooth.h and modbluetooth_btstack.c)

jonathanhogg replied:
This is returning the errno from the stack, though, same as the other `mp_bluetooth_` methods.

jimmo replied:
Sorry, I mean

```c
int mp_bluetooth_gap_peripheral_connect_cancel(void);
```

Otherwise C treats "void foo();" as a function that takes any number of arguments.

jonathanhogg replied:
Oops! Yes, of course, sorry! I'm slightly embarrassed to admit that my C stretches back to K&R and I can be a bit slapdash about the "modern" conventions ;-)

jimmo reviewed (commented):
Thanks @jonathanhogg, this looks useful. I will test this on STM32 btstack+nimble and update.

jimmo commented:
Maybe just one thought... instead of adding a new method, would `ble.gap_connect(None)` be sufficient? (Just to save a few tens of bytes...)

jonathanhogg commented:
Hey @jimmo, good call! Yes, using `ble.gap_connect(None)` would likely match the existing API more closely. I'll hammer in that change.

Interestingly, this doesn't fix my problem anyway. It turns out that the stuck connections I am suffering from are actually a result of missing events from the BLE IRQ callback. I'm about to raise an issue to capture thoughts.

jimmo commented on extmod/modbluetooth.c:
```diff
@@ -630,6 +630,13 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_2(bluetooth_ble_gatts_register_services_obj, blue
+            int err = mp_bluetooth_gap_peripheral_connect_cancel();
+            return bluetooth_handle_errno(err);
+        }
+        mp_raise_TypeError(MP_ERROR_TEXT("None or addr_type and addr required"));
```
Can we make this also "invalid addr" (so it uses the same string as the error message below).

jonathanhogg replied:
Can do!

jimmo reviewed (commented):
Other than the error message this looks good to me. Thanks @jonathanhogg !

jonathanhogg committed:
extmod/modbluetooth: Support gap_connect(None) to cancel a connection.

Allow cancellation of in-progress peripheral connections.

jonathanhogg committed:
docs/library/ubluetooth: Document cancellation of peripheral connect.

Document new `gap_connect(None)` mechanism for cancelling an in-progress 
connection attempt.

dpgeorge commented:
Thanks for this, now merged in 851ecb2da178fff0b60aefdb5af502f28787a7ec
